pinot-core/src/test/java/org/apache/pinot/core/common/datatable/DataTableSerDeTest.java

```diff
@@ -163,6 +163,135 @@ public void testAllDataTypes()
     verifyDataIsSame(newDataTable, columnDataTypes, numColumns);
   }
 
+  @Test
+  public void testV3V4Compatibility()
+      throws IOException {
+    DataSchema.ColumnDataType[] columnDataTypes = DataSchema.ColumnDataType.values();
+    int numColumns = columnDataTypes.length;
+    String[] columnNames = new String[numColumns];
+    for (int i = 0; i < numColumns; i++) {
+      columnNames[i] = columnDataTypes[i].name();
+    }
+
+    DataSchema dataSchema = new DataSchema(columnNames, columnDataTypes);
+
+    // Verify V4 broker can deserialize data table (has data, but has no metadata) send by V3 server
```

Comment:
If not already covered, can we write a test to simulate the following scenario ?

- same cluster (broker tenant and server tenant) are processing 2 queries 
-- 1 query on the existing engine
-- 1 query on the multi-stage engine
- both broker and server have been upgrade to new version that has V4
- for query 1, server sends a data table that has no concept of row / columnar / metadata blok / block type
- for query 2, server sends a data table that is aware of multi-engine's DataBlock
- Broker is able to correctly handle both data tables

Reply:
multi-stage engine is not being considered at this moment. because there's a issue with 
DataTableV4 is needed for multi-stage engine
DataTalbeV3 is still default in normal engine. 

Since the `_version` member in DataTableBuilder is static, we cant make the dynamic switch. 
In order to address this we will have to make the `_version` as part of the constructor parameter and no longer make the builder static. 

I will look into how this can be done but since multi-stage engine is not yet prime time. I think we can stick with just the one static version per component approach

Reply:
Add a TODO ?